### PR TITLE
✨🔧 452 - [Feature branch] - Add ega permissions integration

### PR DESCRIPTION
**Pull request by anncatton**
**Note**: This will be the base branch for work associated with the new EGA permissions flow.

Includes work from:
- https://github.com/icgc-argo/dac-api/pull/462
- https://github.com/icgc-argo/dac-api/pull/460
- https://github.com/icgc-argo/dac-api/pull/458

Adds initial setup for EGA API client.

## EGA Job
- updates `config.ts` to include values for EGA APIs
- adds axios clients for EGA auth server and API
- creates client for interacting with EGA API, including fetching an access token from auth server
- adds initial logic for refreshing the access token, using Axios interceptors. Will need to expand this to address multiple concurrent requests!
- adds constants for API paths

## Documentation
- adds a new section for environment variables in the README. For now, just added the vars related to EGA. Will add remaining values in a cleanup ticket, as some of them are likely to be removed (i.e. Vault values)
- updates `.env.example` file

## Docker Compose
- adds specific version for Vault image, there is no `latest` image available on dockerhub
- changes `mailhog` image to a forked version to allow building Multi Arch Docker images: https://hub.docker.com/r/jcalonso/mailhog
- modifies Makefile scripts to match updated Docker executable

## Dependencies
- adds `axios` and types support for `jsonwebtoken`

## Cleanup
- unrelated to ticket, moves some utils functions from the router file (`applications.ts`) to their own file

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -48,34 +48,34 @@ DACO_REVIEW_POLICY_NAME=DACO-REVIEW
...
+EMAIL_REVIEWER_FIRSTNAME=DACO
+EMAIL_REVIEWER_LASTNAME=ADMIN
 DCC_MAILING_LIST=
 DACO_SURVEY_URL=
 
 ##############
 # UI         #
 ##############
-DACO_UI_BASE_URL=https://dac.dev.argo.cancercollaboratory.org
+DACO_UI_BASE_URL=http://localhost:3000
```
Comment (joneubank) [inline .env.example:78]: Thank you for removing specific system references.

File: README.md [hunk 2]
```diff
@@ -8,6 +8,17 @@ Development of the Data Access Control API
...
+
+| Name                | Description                                                                   | Type     | Required | Default |
+| ------------------- | ----------------------------------------------------------------------------- | -------- | -------- | ------- |
+| EGA_CLIENT_ID       | Client ID for EGA API                                                         | `string` | true     |         |
+| EGA_AUTH_HOST       | Root URL for EGA authentication server                                        | `string` | true     |         |
+| EGA_AUTH_REALM_NAME | Realm name for EGA authentication server                                      | `string` | true     |         |
+| EGA_API_URL         | Root URL for EGA API                                                          | `string` | true     |         |
+| EGA_USERNAME        | Username for account used to gain access token from EGA authentication server | `string` | true     |         |
+| EGA_PASSWORD        | Password for account used to gain access token from EGA authentication server | `string` | true     |         |
+
```
Comment (joneubank) [inline README.md:21]: 🎉

File: src/jobs/ega/egaClient.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      password: egaPassword,
+    },
+    {
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+    },
+  );
+
+  const token = response.data;
```
Comment (joneubank) [inline src/jobs/ega/egaClient.ts:91]: `data` has type `any`. Are we planning to ensure that the response data matches the type `IdpToken`?

File: src/jobs/ega/egaClient.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      refresh_token: token.refresh_token,
+    },
+    {
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+    },
+  );
+
+  return response.data;
```
Comment (joneubank) [inline src/jobs/ega/egaClient.ts:114]: Repeat comment, `data` has type `any`.

File: src/routes/utils.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import { BadRequest } from '../utils/errors';
+
+export function validateId(id: string) {
```
Comment (joneubank) [inline src/routes/utils.ts:22]: TSDocs please, no idea what this is validating or when to use.

**Review by joneubank — CHANGES_REQUESTED** [review]
Comments are about code strictness and readability/maintainability. There's nothing very functional happening here to request changes for.

File: src/utils/constants.ts [hunk 6]
```diff
@@ -271,3 +271,19 @@ export const ICGC_ARGO_CONTACT_URL = urlJoin(ICGC_ARGO_PLATFORM_URL, 'contact');
...
+// ega api
+
+export const EGA_API = {
+  DACS: 'dacs',
+  PERMISSIONS: 'permissions',
+  USERS: 'users',
+  MEMBERS: 'members',
+  REQUESTS: 'requests',
+  DATASETS: 'datasets',
+};
```
[the remark this answers is not in the document]
Comment (anncatton) [reply src/utils/constants.ts:289]: these are pathnames for the ega api requests, they'll come into play in the next PR(s). I'll add a comment tho

File: src/jobs/ega/egaClient.ts (hunk 3 above)
Comment (anncatton) [reply src/jobs/ega/egaClient.ts:91]: i'm adding Zod in the [next PR,](https://github.com/icgc-argo/dac-api/pull/458) i will add a `safeParse` call for this and the refreshToken return

File: src/jobs/ega/egaClient.ts (hunk 4 above)
Comment (anncatton) [reply src/jobs/ega/egaClient.ts:114]: Will fix as in above comment

**Review by joneubank — APPROVED** [review]

**Comment by joneubank** [comment]
Before you merge this, you mentioned "Next PR" a couple times, should we be collecting this work into a larger feature branch?

**Comment by anncatton** [comment]
> Before you merge this, you mentioned "Next PR" a couple times, should we be collecting this work into a larger feature branch?

My plan was to use this branch as the feature branch - the "next PR" is the work for #453 https://github.com/icgc-argo/dac-api/pull/458, which uses this as its base branch.

**Pull request merged**